models/dps/writer.go

```diff
@@ -36,4 +36,6 @@ type Writer interface {
 	Transactions(height uint64, transactions []*flow.TransactionBody) error
 	Results(results []*flow.TransactionResult) error
 	Seals(height uint64, seals []*flow.Seal) error
+
+	Close() error
```

Comment:
This should not be added to the interface. Better to find another solution then, on how to initialize in the main file.

Reply:
The obvious solution I see is to no longer wrap the writer but compose it instead, but then the issue becomes that the writer needs to either have some logic such as:

```go
// Header indexes the given header of a finalized block at the given height.
func (w *Writer) Header(height uint64, header *flow.Header) error {
	if cfg.MetricsEnabled {
		w.metrics.Header()
	}
	return w.apply(w.lib.SaveHeader(height, header))
}
```

Or to have a noopCounter in the case where we want to disable metrics, but still always call it:

```go
// Header indexes the given header of a finalized block at the given height.
func (w *Writer) Header(height uint64, header *flow.Header) error {
	w.metrics.Header()
	return w.apply(w.lib.SaveHeader(height, header))
}
```

Unfortunately I don't like both of these solutions 🤔 I'm looking for some other way to design this right now.

Reply:
I didn't find any alternatives that aren't overcomplicated so I went with one of these above solutions. Let me know if you have something in mind that would be better. I wanted to keep the writer completely unaware of anything metrics-related so this isn't great, but it makes the `main` cleaner and clearer.

Having an option for enabling metrics as opposed to a wrapper makes it easier to understand for readers I assume. It just sucks that there's extra logic in the writer now.

One other way that I thought of would be to always wrap it in a metrics counter, but make it not count anything and just directly call the writer's methods if metrics are disabled, but I'm not crazy about it either because then we're basically wrapping the writer in something completely useless. Also this would mean we'd still need to have the `Close()` method in the interface.

LMK if you have some other ideas 🤷

Reply:
Actually I just realized, this solution is wrong too anyway, because this means we can't instantiate more than one index.Writer at any given time (otherwise we trigger a panic due to registering the same counters multiple times) which breaks the integration tests. I'll revert this change for now and get back to it later, hopefully with a fresh idea.

Reply:
Another solution is to have both the writer and metricsWriter in the command, and when we pass the writer, check for whether metrics are enabled, and if they are, pass the metricsWriter instead. Not very elegant but keeps the rest of the code very clean. The only place which isn't great is the command's main. I'll do this for now actually.